### DISCUSSION TITLE
Function not found in customer PropertyBag

### DISCUSSION THREAD

**Discussion by ZelgiusLiu** [ClearFoundry/ClearScript#522] [Q&A] [answered]
Hello, I am using ClearScript and now I have a problem.
I tried to implement a class extending IPropertyBag.
For using indexer, it works fine now, but I found that when I want to call functions, e.g. ```myClass.testFunction()```, it threw an exception, said "myClass.testFunction is not a function."
So, how can I get access to this function? Or how can I use custom functions in a extended class for IPropertyBag?
Thanks.

**Comment by ClearScriptLib** [comment]
Hi @ZelgiusLiu,

How are you adding or defining `testFunction`?

Thanks!

**Reply by ZelgiusLiu** [reply]
Certainly, I defined this method in my class, like below:
```CSharp
public class MyValues : IPropertyBag
{
        Dictionary<string, object> values;

        public MyValues()
        {
            values = new Dictionary<string, object>();
        }

        #region Implement For IDictionary<string, object>
        ......
        #endregion

        public void Print()
        {
            Console.WriteLine(values.Count);
        }
}
```
And then, I execute code like this:
```CSharp
          using var engine = new V8ScriptEngine();
          engine.AddHostObject("values", new MyValues()
          {
              ["greeting"] = "Hello world",
          });
          engine.Execute("values.Print();");
```
While executing this code, it threw the error: "TypeError: values.Print is not a function".
Is there any mistake I made in the code?

**Reply by ClearScriptLib** [reply, marked as answer]
Hi @ZelgiusLiu,

From the [`IPropertyBag`](https://clearscript.clearfoundry.net/Reference/html/T_Microsoft_ClearScript_IPropertyBag.htm) documentation (emphasis added):

> If an object that implements this interface is added to a script engine (see [`AddHostObject`](https://clearscript.clearfoundry.net/Reference/html/M_Microsoft_ClearScript_ScriptEngine_AddHostObject_1.htm)), script code will be able to access the properties stored in the collection as if they were members of the object itself, using the script language's native syntax for member access. **No other members of the object will be accessible.**

The idea is to eliminate potential ambiguity. In C#, indexing and member access are distinct operations, whereas in JavaScript they're exactly the same – e.g., `values['Print']` is 100% equivalent to `values.Print` – and ClearScript has no way to determine whether you're trying to access a member named "Print" or an item named "Print" in the name-value collection. `IPropertyBag` avoids the conflict by exposing only the name-value pairs to script code.

However, nothing stops you from adding an item whose value is a delegate that invokes the desired method:

```C#
public MyValues() {
    values = new Dictionary<string, object> {
        ["Print"] = new Action(Print)
    };
}
```

Another possibility is to use a second internal collection to hold delegates, to be exposed only when the corresponding keys in the main collection are unpopulated:

```C#
private readonly Dictionary<string, object> delegates;
public MyValues() {
    values = new Dictionary<string, object>();
    delegates = new Dictionary<string, object> {
        ["Print"] = new Action(Print)
    };
}
...
public bool TryGetValue(string key, out object value) {
    return values.TryGetValue(key, out value) || delegates.TryGetValue(key, out value);
}
```

Or you could customize `TryGetValue` further to get the exact behavior you require.

Good luck!

**Reply by ZelgiusLiu** [reply]
I tried with your method and it shall work.
Thx for the idea.
